client/rest/src/test/java/org/elasticsearch/client/RestClientTests.java

```diff
@@ -48,50 +51,83 @@ public void testCloseIsIdempotent() throws IOException {
     }
 
     public void testPerformAsyncWithUnsupportedMethod() throws Exception {
-        RestClient.SyncResponseListener listener = new RestClient.SyncResponseListener(10000);
+        final CountDownLatch latch = new CountDownLatch(1);
         try (RestClient restClient = createRestClient()) {
-            restClient.performRequestAsync("unsupported", randomAsciiLettersOfLength(5), listener);
-            listener.get();
-
-            fail("should have failed because of unsupported method");
-        } catch (UnsupportedOperationException exception) {
-            assertEquals("http method not supported: unsupported", exception.getMessage());
+            restClient.performRequestAsync("unsupported", randomAsciiLettersOfLength(5), new ResponseListener() {
+                @Override
+                public void onSuccess(Response response) {
+                    fail("should have failed because of unsupported method");
+                }
+
+                @Override
+                public void onFailure(Exception exception) {
+                    assertThat(exception, instanceOf(UnsupportedOperationException.class));
+                    assertEquals("http method not supported: unsupported", exception.getMessage());
```

Comment:
off topic : I see that both `assertThat` and `assertEquals` are used throughout the tests. Is there a preference to use one over the other ( for new code ) ?

Reply:
Depends on who you ask. I prefer to use the purpose built assertions when they are the same as the ones from Hamcrest because they are fairly terse and don't need to make objects [1]. I prefer to use the hamcrest ones when they offer better error reporting. I rarely go out of my way to write new hamcrest matchers. By that logic, I could have written `assertThat(e, hasMessage(equalTo("http method not supported: unsupported")));` then I would have done so. But `hasMessage` isn't a thing.

Other folks always use `assertThat` because has very obvious parameter ordering that reads well and the compile can check it. This is a totally valid argument and one that I'll probably end up being swayed by eventually.

I tend not to rewrite assertions that someone else wrote unless I have a good reason to.

[1]: I imagine hamcrest matchers are a great candidate for the JVM to optimize to stack based allocation but I figure that only kicks in for hot method and most test methods aren't run enough to be hot.